thinc/util.py

```diff
@@ -172,19 +123,18 @@ def to_numpy(data):  # pragma: no cover
         return numpy.array(data)
 
 
-def set_active_gpu(gpu_id: int) -> "cupy.cuda.Device":  # pragma: no cover
+def set_active_gpu(gpu_id: int) -> Optional["cupy.cuda.Device"]:  # pragma: no cover
     """Set the current GPU device for cupy and torch (if available)."""
-    import cupy.cuda.device
+    if not gpu_is_available():
+        return None
 
     device = cupy.cuda.device.Device(gpu_id)
     device.use()
-    try:
-        import torch
 
+    if has_torch_gpu:
         torch.cuda.set_device(gpu_id)
         torch.set_default_tensor_type("torch.cuda.FloatTensor")
-    except ImportError:
-        pass
+
     return device
```

Comment:
I worry that this might be slightly breaking behaviour, too. So before we would throw an error if you'd request to set the active GPU but that failed, and now we'd just silently move on?

Reply:
I guess it depends on how we want to view the original behaviour. Previously, we were importing `cupy` without the necessary sanity checks (like we do elsewhere), opening up the possibility of an unhandled exception. This wasn't specified in the function's documentation (and the "if available" phrase alludes to some amount of exception safety, I think), so I interpreted this lack of a sanity check as a bug.

That said, I now see that the  "if available" phrase could parsed as an adjunct to _just_ `torch`, in which case it would in fact make sense to not modify the function behaviour per-se but to explicitly raise an exception before we call into `cupy`; the `cupy` call will eventually throw under such circumstances anyway). I take that this would be preferable?

Reply:
I read it as the "if available" being applicable only to the "torch" bit, but it's definitely ambiguous. 

Raising an exception would make sense to me. The docs provide little additional clarification (let's add some) but the current typing to `cupy.cuda.Device` (without `Optional`) signals to me that we'd always like this to work, and if we can't get a proper device, the function should raise an exception.